src/App.tsx

```diff
@@ -1,18 +1,39 @@
-import React from 'react';
+import React, { useState } from 'react';
 import './App.css';
 import { getNumbers } from './utils';
+import { Pagination } from './components/Pagination';
 
 // eslint-disable-next-line @typescript-eslint/no-unused-vars
 const items = getNumbers(1, 42)
   .map(n => `Item ${n}`);
 
 export const App: React.FC = () => {
+  const total = 42;
+  const [perPage, setPerPage] = useState(5);
+  const [currentPage, setCurrentPage] = useState(1);
+
+  const onPageChange = (page: number) => (
+    setCurrentPage(page)
+  );
+
+  const handleChange = (event: React.ChangeEvent<HTMLSelectElement>) => {
+    event.preventDefault();
+    setPerPage(+event.target.value);
+    setCurrentPage(1);
+  };
+
+  const lastItem = currentPage * perPage > total
+    ? total
+    : currentPage * perPage;
+  const firstItem = (currentPage * perPage) - perPage + 1;
+  const itemsToShow = items.slice(firstItem - 1, lastItem);
```

Comment:
you need to use hook useMemo here, because every time when you state was changed method slice() will be called
so it's not okay for performance